cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/MetadataClient.java

```diff
@@ -4,123 +4,164 @@
-        long start = _time.milliseconds();
-        _networkClient.poll(remaining, start);
-        remaining -= (_time.milliseconds() - start);
-        isMetadataUpdated = _metadata.version() > version;
+    if (_time.milliseconds() >= _lastSuccessfulUpdate + _metadataTTL) {
+      // Cruise Control always fetches metadata for all the topics.
+      boolean isMetadataUpdated = false;
+      Optional<Cluster> refreshedCluster = Optional.empty();
+      try {
+        refreshedCluster = Optional.of(doRefreshMetadata(timeout));
+        isMetadataUpdated = true;
+      } catch (ExecutionException | TimeoutException | InterruptedException e) {
+        LOG.warn("Exception while updating metadata", e);
       }
-      if (isMetadataUpdated) {
+
+      if (isMetadataUpdated && refreshedCluster.isPresent()) {
+        _lastSuccessfulUpdate = _time.milliseconds();
+        _version += 1;
         if (LOG.isDebugEnabled()) {
-          LOG.debug("Updated metadata {}", _metadata.fetch());
+          LOG.debug("Updated metadata {}", _cluster);
         }
-        if (MonitorUtils.metadataChanged(beforeUpdate, _metadata.fetch())) {
+        Cluster newCluster = refreshedCluster.get();
+        if (MonitorUtils.metadataChanged(_cluster, newCluster)) {
           _metadataGeneration.incrementAndGet();
+          _cluster = newCluster;
         }
       } else {
         LOG.warn("Failed to update metadata in {}ms. Using old metadata with version {} and last successful update {}.",
-                 timeout, _metadata.version(), _metadata.lastSuccessfulUpdate());
+                 timeout, _version, _lastSuccessfulUpdate);
       }
     }
-    return new ClusterAndGeneration(_metadata.fetch(), _metadataGeneration.get());
+    return new ClusterAndGeneration(_cluster, _metadataGeneration.get());
   }
 
-  /**
-   * Close the metadata client. Synchronized to avoid interrupting the network client during a poll.
-   */
-  public synchronized void close() {
-    _networkClient.close();
+  private Cluster doRefreshMetadata(int timeout) throws InterruptedException, ExecutionException, TimeoutException {
+    // We use both AdminClient timeout options and `future.get(timeout)`
+    // The former ensures the timeouts are enforced in the brokers
+    int remaining = timeout;
+    long start = _time.milliseconds();
+
+    Set<String> topicNames = _adminClient.listTopics(new ListTopicsOptions().timeoutMs(remaining))
+        .names().get(remaining, TimeUnit.MILLISECONDS);
+    long end = _time.milliseconds();
+    long elapsedTimeMs = end - start;
+    remaining -= elapsedTimeMs;
+
+    DescribeClusterResult result = _adminClient.describeCluster(new DescribeClusterOptions().timeoutMs(remaining));
+    Collection<Node> nodes = result.nodes().get(remaining, TimeUnit.MILLISECONDS);
+    String clusterId = result.clusterId().get(remaining, TimeUnit.MILLISECONDS);
+    Node controller = result.controller().get(remaining, TimeUnit.MILLISECONDS);
```

Comment:
(Similar to the issue above) Remaining timeout is not adjusted properly between get calls to future.

Reply:
I think the same thing applies here - `_adminClient.describeCluster(new DescribeClusterOptions().timeoutMs(remaining))` simply instructs the background thread to give up after `remaining`.

The consequent three `get(remaining, TimeUnit.MILLISECONDS)` calls do not adjust the `remaining` because they rely on the fact that only the first get will take any time - the others should be instant because the data has been fetched